### PR TITLE
Fix empty states for spaces

### LINKED ISSUE #3327: #+ button on lower right when looking at an empty space goes to an empty 'Explore rooms'

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes #3327

image

Also added quick menu to add existing room from explore if you are admin
image

As well as empty item for add room screen
image

### Pull Request Checklist

- [x] Changes has been tested on an Android device or Android emulator with API 21
- [x] UI change has been tested on both light and dark themes
- [x] Pull request is based on the develop branch
- [x] Pull request updates [CHANGES.md](https://github.com/vector-im/element-android/blob/develop/CHANGES.md)
- [x] Pull request includes screenshots or videos if containing UI changes
- [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)

### CODE REVIEW

**Comment by BillCarsonFr** [comment]
@niquewoodhouse can you review the copy here?

> This space has no rooms

> Some rooms may be hidden because they’re private and you need an invite.\nYou don’t have permission to add rooms.

We are saying has no rooms, but it could also be no sub spaces.
Should we talk about rooms and sub spaces, or just states that it's empty to you ?

File: vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryController.kt [hunk 1]
```diff
@@ -96,9 +98,23 @@ class SpaceDirectoryController @Inject constructor(
...
             if (flattenChildInfo.isEmpty()) {
-                genericFooterItem {
-                    id("empty_footer")
-                    stringProvider.getString(R.string.no_result_placeholder)
+                genericEmptyWithActionItem {
+                    id("empty_res")
+                    title(stringProvider.getString(R.string.this_space_has_no_rooms))
+                    iconRes(R.drawable.ic_empty_icon_room)
+                    iconTint(colorProvider.getColorFromAttribute(R.attr.riotx_reaction_background_on))
+                    apply {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryController.kt:106]: I do not think using a `apply` block is necessary here.

File: vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryFragment.kt [hunk 2]
```diff
@@ -77,6 +88,32 @@ class SpaceDirectoryFragment @Inject constructor(
         views.toolbar.title = title
     }
 
+    override fun onPrepareOptionsMenu(menu: Menu) = withState(viewModel) { state ->
+        menu.findItem(R.id.spaceAddRoom)?.let {
+            it.isVisible = state.canAddRooms
+        }
+        menu.findItem(R.id.spaceCreateRoom)?.let {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryFragment.kt:95]: no need to use `let` blocks here, you can simply write
`menu.findItem(R.id.spaceAddRoom)?.isVisible = state.canAddRooms`

File: vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryFragment.kt [hunk 3]
```diff
@@ -77,6 +88,32 @@ class SpaceDirectoryFragment @Inject constructor(
...
+            it.isVisible = false // Not yet implemented
+        }
+        super.onPrepareOptionsMenu(menu)
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        when (item.itemId) {
+            R.id.spaceAddRoom -> {
+                withState(viewModel) { state ->
+                    addExistingRooms(state.spaceId)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryFragment.kt:105]: I would have let `addExistingRooms` to retrive the spaceId from the state, but OK

**Review by bmarty — APPROVED** [review]
LGTM, just a few remarks

File: vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryController.kt (hunk 1 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryController.kt:106]: done

File: vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryFragment.kt (hunk 2 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/app/features/spaces/explore/SpaceDirectoryFragment.kt:95]: done

**Pull request merged**